### ISSUE TITLE
Masking Validation Function for Testing tokenlist file

### ISSUE THREAD

**Issue by sigkill** [3rdIteration/btcrecover#238]
I haven't been able to find a good way to print out the output of a mask - have been working on adding some switches to add hashcat type mask conversion support, but I was curious before getting too deep if there is a way to output to stdout what my mask is actually sending to the password checker?

For instance if I were to create a mask in hashcat:
?u?u?u?d

`Built-in charsets
?l = abcdefghijklmnopqrstuvwxyz
?u = ABCDEFGHIJKLMNOPQRSTUVWXYZ
?d = 0123456789
?h = 0123456789abcdef
?H = 0123456789ABCDEF
?s = «space»!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
?a = ?l?u?d?s
?b = 0x00 - 0xff`

But in btcrecover it's a bit weird because some of the examples include inputing characters where you might have part of a password.

It's not clear to me if the separator character settings are also taking effect - like if I have a particular character I'd like a way to test to see what btcrecover.py is generating from the token / masks.

Is there a way to load this function into python and check? For example: a password like "35!leaf!a3c2!XRt2!66" with a clear separator character of "!" and no other special characters - I'm trying to find out if my mask is working properly, but haven't found a good way to verify it, but have started working on a crappy sed script to convert hashcat / pack to btcrecover. Another really handy tool I found was the hashcat mask processor - which reads a given mask and helps generate a dictionary, or lets you use some of the mask functions from hashcat into btcrecover.py

Given the above mask / token string I think I should be able the create a token like ^35!%l%l%l%l!%l%d%l%d!%A%A%l%d!%d%d

Would this be correct?

Anyhow any help for mask validation would be very helpful as I am stuck for the moment.

**Comment by sigkill** [comment]
Would using %N be better?

**Comment by sigkill** [comment]
Okay! Found how to do it after going through the code on the weekend:

`python3  ./btcrecover.py  --tokenlist ./testtoken --dsw --no-eta   --no-dupchecks  --custom-wild !  --threads 8 --listpass`

If I use a tokenlist, it effectively becomes a password generator with the string I've setup, and lets me validate the passwords that are being generated off of my token string!

A few thoughts - nice to have would be a step sequencer - something that can tell you the first, and last passwords it would try, and total passwords between that. If this were offloaded to a gpu / opencl function that could then break up the steps between the very first and very last sequences.

Also the bip38 function in python is horrendously slow, but does not have to be. I'm going to update the other issue I opened, or create a new issue for that. I compared with using wallycore and the other bip38 project I mentioned in that comment. That project is in C++ and does close to 2000 passwords per second.
